### PR TITLE
Add Torch export for HSTU model

### PR DESCRIPTION
**Pull request by jensenhwa**

### CODE REVIEW

File: examples/hstu/inference/inference_gr_ranking.py [hunk 1]
```diff
@@ -447,4 +651,9 @@ def strip_padding_batch(batch, unpadded_batch_size):
             disable_kvcache=args.disable_kvcache,
             max_bs=args.max_bs,
         )
+    elif args.mode == RunningMode.SNAP:
```
Comment (shijieliu) [inline examples/hstu/inference/inference_gr_ranking.py]: Adding a new mode for export is strange to me. Can we replace the inference code path w/o kvcache with this new exportable implementation?

File: examples/hstu/inference/README.md [hunk 2]
```diff
@@ -73,6 +73,23 @@ Turn on option `INFERENCEBUILD=1` to skip Megatron installation, which is not re
 ~$ python3 ./inference/inference_gr_ranking.py --gin_config_file ./inference/configs/kuairand_1k_inference_ranking.gin --checkpoint_dir ${PATH_TO_CHECKPOINT}  --mode eval
 ```
 
+## Example: Kuairand-1K for Snap
```
Comment (shijieliu) [inline examples/hstu/inference/README.md]: it's better for us not to name it as Snap

File: examples/hstu/modules/inference_dense_module.py
Comment (shijieliu) [inline examples/hstu/modules/inference_dense_module.py]: choose between HSTUBlockInference or HSTUBlock by configuration @geoffreyQiu

File: docker/Dockerfile.torch211 [hunk 3]
```diff
@@ -0,0 +1,103 @@
+ARG BASE_IMAGE=nvcr.io/nvidia/pytorch:26.02-py3
```
Comment (shijieliu) [inline docker/Dockerfile.torch211]: @geoffreyQiu try merge this dockerfile with existing dockerfile. if needed, we can update torchrec to 1.4.0 or 1.5.0. but torchrec 1.5.0 needs compatibility change in dynamicemb

File: third_party/FBGEMM
Comment (shijieliu) [inline third_party/FBGEMM:1]: merge this into fbgemm dev after verified

File: examples/hstu/inference/inference_gr_ranking.py (hunk 1 above)
Comment (geoffreyQiu) [reply examples/hstu/inference/inference_gr_ranking.py]: Removed this section, and split the torch export process into another file.

File: examples/hstu/inference/README.md (hunk 2 above)
Comment (geoffreyQiu) [reply examples/hstu/inference/README.md]: Removed.

File: examples/hstu/modules/inference_dense_module.py
Comment (geoffreyQiu) [reply examples/hstu/modules/inference_dense_module.py]: Done. [link](https://github.com/jensenhwa/recsys-examples/blob/jhwa/snap_export/examples/hstu/modules/inference_dense_module.py#L142)

File: docker/Dockerfile.torch211 (hunk 3 above)
Comment (geoffreyQiu) [reply docker/Dockerfile.torch211]: Removed this file, as we have upgraded to Torch 2.11 + CUDA 13.1 (Pytorch:26.02) in #347 .

File: third_party/FBGEMM
Comment (geoffreyQiu) [reply third_party/FBGEMM:1]: Updated as in #347.

File: examples/commons/ops/triton_ops/triton_jagged.py [hunk 4]
```diff
@@ -1129,15 +1129,15 @@ def forward(
...
             if seq_len_a is None:
-                seq_len_a = int(offsets_a[-1].item())
+                seq_len_a = offsets_a[-1].item()
             if seq_len_b is None:
-                seq_len_b = int(offsets_b[-1].item())
+                seq_len_b = offsets_b[-1].item()
         _, D = values.shape
         BLOCK_D = triton.next_power_of_2(D)
         values_a = torch.empty((seq_len_a, D), device=values.device, dtype=values.dtype)
         values_b = torch.empty((seq_len_b, D), device=values.device, dtype=values.dtype)
```
[the remark this answers is not in the document]
Comment (geoffreyQiu) [reply examples/commons/ops/triton_ops/triton_jagged.py:1138]: `.item()` is avoided by having `seq_len_a` or `seq_len_b` is never `None`.

File: examples/hstu/modules/hstu_processor.py [hunk 5]
```diff
@@ -208,11 +225,12 @@ def hstu_preprocess_embeddings(
...
+            total_candidates_seq_len = num_candidates.sum()
         elif contextual_seqlen is not None:
-            total_candidates_seq_len = int(
-                (sequence_embeddings_lengths.sum() - contextual_seqlen.sum()).item()
-            )
+            total_candidates_seq_len = sequence_embeddings_lengths.sum() - contextual_seqlen.sum()
+    elif torch.compiler.is_compiling():
+        assert num_candidates is not None, "num_candidates should be None during inference when compiling"
+        total_candidates_seq_len = num_candidates.sum()
     return JaggedData(
```
[the remark this answers is not in the document]
Comment (geoffreyQiu) [reply examples/hstu/modules/hstu_processor.py:234]: We assume no torch export when `is_inference` is `True`. Skip this change for now.

File: examples/hstu/inference/export_inference_gr_ranking.py [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+        dataloader = get_data_loader(dataset=eval_dataset)
+        dataloader_iter = iter(dataloader)
+
+        def strip_padding_batch(b):
+            b = b.to(device=torch.cuda.current_device())
+            d = b.features.to_dict()
+            user_ids = d["user_id"].values().cpu().long()
+            if user_ids.shape[0] != b.batch_size:
+                b = strip_padding_batch(b, user_ids.shape[0])
+            return b
```
[the remark this answers is not in the document]
Comment (geoffreyQiu) [reply examples/hstu/inference/export_inference_gr_ranking.py:373]: Done.

File: examples/hstu/inference/export_inference_gr_ranking.py [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+
+        sc[batch.features.values()] = {0: Dim.AUTO}
+        sc[batch.features.lengths()] = {0: Dim.AUTO}
+        sc[batch.num_candidates] = {0: Dim.AUTO}
+        dynamic_shapes = sc.dynamic_shapes(model, (batch,))
+        print(f"[INFO] Dynamic shapes: {dynamic_shapes}")
+
+        if debug_flattened_inputs:
+            debug_print_flattened_export_args(batch, embeddings)
+
```
[the remark this answers is not in the document]
Comment (geoffreyQiu) [reply examples/hstu/inference/export_inference_gr_ranking.py:405]: Done

File: examples/hstu/inference/cpp_inference/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+target_link_libraries(inferece_hstu_gr_ranking_exported_model PRIVATE hstu_cuda_ops_runtime)
+target_link_libraries(inferece_hstu_gr_ranking_exported_model PRIVATE ${CMAKE_DL_LIBS})
+target_link_libraries(inferece_hstu_gr_ranking_exported_model PRIVATE ${NVE_TORCH_OPS_LIB} ${NVE_COMMON_LIB})
+target_compile_definitions(inferece_hstu_gr_ranking_exported_model PRIVATE HSTU_CUDA_OPS_AVAILABLE=1)
+
+target_include_directories(inferece_hstu_gr_ranking_exported_model PRIVATE
+    ${TORCH_INCLUDE_DIRS}
+    /home/junyiq/newscratch/spring26/trt-recsys/
+    /home/junyiq/newscratch/spring26/trt-recsys/include
+)
```
[the remark this answers is not in the document]
Comment (geoffreyQiu) [reply examples/hstu/inference/cpp_inference/CMakeLists.txt]: Done.

File: examples/hstu/inference/cpp_inference/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,107 @@
+cmake_minimum_required(VERSION 3.18)
+project(inferece_hstu_gr_ranking_exported_model LANGUAGES CXX CUDA)
+
```
[the remark this answers is not in the document]
Comment (geoffreyQiu) [reply examples/hstu/inference/cpp_inference/CMakeLists.txt:3]: Done.

File: examples/commons/ops/cuda_ops/csrc/kjt_aux_op.cpp [hunk 10]
```diff
@@ -0,0 +1,157 @@
+#include <ATen/ATen.h>
+#include <torch/library.h>
+
+namespace {
+
+std::vector<at::Tensor> split_by_lengths_impl(
+    const at::Tensor& values,
+    const at::Tensor& lengths_1d,
+    int64_t num_splits) {
```
[the remark this answers is not in the document]
Comment (geoffreyQiu) [reply examples/commons/ops/cuda_ops/csrc/kjt_aux_op.cpp:10]: Done.

File: examples/hstu/modules/exportable_embedding.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+
+
+_AOTI_DEMO_LIB_DIR = os.path.join(os.path.dirname(__file__), "aoti_demo", "lib")
+
+_SEARCH_PATHS = [
+    os.path.join(_AOTI_DEMO_LIB_DIR, "inference_emb_ops.so"),
+    os.path.join("aoti_demo", "lib", "inference_emb_ops.so"),
+    os.path.join("inference", "aoti_demo", "lib", "inference_emb_ops.so"),
+    "inference_emb_ops.so",
+]
```
[the remark this answers is not in the document]
Comment (geoffreyQiu) [reply examples/hstu/modules/exportable_embedding.py]: Done.

File: examples/hstu/modules/exportable_embedding.py [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+                last_index += 1
+        self.num_features = len(self.feature_to_index)
+        
+        self._embedding_table = InferenceEmbeddingTable(
+            self.table_options,
+            self.pooling_mode,
+            self.table_names,
+            self.feature_table_map,
+            device=torch.device("cuda:0"),
+        )
```
[the remark this answers is not in the document]
Comment (geoffreyQiu) [reply examples/hstu/modules/exportable_embedding.py:134]: Done.

File: corelib/dynamicemb/dynamicemb/exportable_tables.py [hunk 13]
```diff
@@ -0,0 +1,502 @@
...
+            - ``pooling_mode_ == -1``: ``(N, D)`` float tensor of per-key
+              embeddings.
+                        - ``pooling_mode_ == 1 or 2``: ``(B, D)`` float tensor of pooled
+              embeddings, where ``B = pooling_offsets.size(0) - 1``.
+        """
+        num_elements = keys.size(0)
+
+        # Derive per-key table id from the table-segment offsets.
+        table_ids = torch.ops.INFERENCE_EMB.expand_table_ids(
+            offsets, keys, None, self.num_tables_, 1, num_elements
```
[the remark this answers is not in the document]
Comment (geoffreyQiu) [reply corelib/dynamicemb/dynamicemb/exportable_tables.py]: Done.

File: examples/hstu/modules/inference_dense_module.py [hunk 14]
```diff
@@ -414,19 +432,33 @@ def forward_nokvcache(
...
 def get_inference_dense_model(
     hstu_config: InferenceHSTUConfig,
     kvcache_config: KVCacheConfig,
     task_config: RankingConfig,
-    use_cudagraph=False,
-    cudagraph_configs=None,
+    use_cudagraph: bool = False,
+    cudagraph_configs: Dict[int, Any] = None,
+    use_exportable: bool = False,
 ):
```
[the remark this answers is not in the document]
Comment (geoffreyQiu) [reply examples/hstu/modules/inference_dense_module.py:456]: Done.

File: corelib/dynamicemb/CMakeLists.txt
Comment (shijieliu) [inline corelib/dynamicemb/CMakeLists.txt:1]: can we update the doc in dynamicemb about how to use inference?

File: examples/hstu/inference/export_inference_gr_ranking.py [hunk 15]
```diff
@@ -0,0 +1,499 @@
...
+        # torch.cuda.profiler.start()
+        while True:
+            try:
+                batch = next(dataloader_iter)
+
+                batch = batch.to(device=torch.cuda.current_device())
+                d = batch.features.to_dict()
+                user_ids = d["user_id"].values().cpu().long()
+                if user_ids.shape[0] != batch.batch_size:
+                    batch = prepare_on_gpu(batch, user_ids.shape[0])
```
[the remark this answers is not in the document]
Comment (geoffreyQiu) [reply examples/hstu/inference/export_inference_gr_ranking.py]: Done.

File: corelib/dynamicemb/dynamicemb/exportable_tables.py [hunk 16]
```diff
@@ -0,0 +1,502 @@
...
+            offsets.append(i)
+            prev = tid
+    offsets.append(len(feature_table_map))
+    return offsets
+
+
+# ---------------------------------------------------------------------------
+# Modules for InferenceEmbeddingTable and its hash table
+# ---------------------------------------------------------------------------
+
```
[the remark this answers is not in the document]
Comment (geoffreyQiu) [reply corelib/dynamicemb/dynamicemb/exportable_tables.py:108]: Done.

File: corelib/dynamicemb/CMakeLists.txt
Comment (geoffreyQiu) [reply corelib/dynamicemb/CMakeLists.txt:1]: Added.

File: examples/hstu/modules/hstu_processor.py [hunk 17]
```diff
@@ -208,11 +225,12 @@ def hstu_preprocess_embeddings(
     total_candidates_seq_len = None
     if not is_inference:
         if num_candidates is not None:
-            total_candidates_seq_len = int(num_candidates.sum().item())
+            total_candidates_seq_len = num_candidates.sum()
         elif contextual_seqlen is not None:
-            total_candidates_seq_len = int(
-                (sequence_embeddings_lengths.sum() - contextual_seqlen.sum()).item()
-            )
+            total_candidates_seq_len = sequence_embeddings_lengths.sum() - contextual_seqlen.sum()
```
[the remark this answers is not in the document]
Comment (geoffreyQiu) [reply examples/hstu/modules/hstu_processor.py:230]: Done.

File: examples/commons/sequence_batch/batch.py [hunk 18]
```diff
@@ -33,7 +33,8 @@ def __post_init__(self):
         if len(set(self.features.keys())) != len(list(self.features.keys())):
             raise ValueError(f"duplicate features keys {list(self.features.keys())}")
         assert isinstance(self.contextual_feature_names, list)
-        assert isinstance(self.batch_size, int) and self.batch_size > 0
+        assert isinstance(self.batch_size, int) and self.batch_size > 0 or \
+            isinstance(self.batch_size, torch.export.dynamic_shapes._IntWrapper) and self.batch_size.val > 0
         self.actual_batch_size = (
```
[the remark this answers is not in the document]
Comment (geoffreyQiu) [reply examples/commons/sequence_batch/batch.py:38]: Use `dynamic_shapes._IntWrapper` for now.

File: examples/commons/datasets/hstu_batch.py [hunk 19]
```diff
@@ -244,7 +244,7 @@ def __post_init__(self):
             self.action_feature_name, str
         ), "action_feature_name must be None or a string"
         assert isinstance(
-            self.max_num_candidates, int
+            self.max_num_candidates, (int, torch.export.dynamic_shapes._IntWrapper)
         ), "max_num_candidates must be an int"
```
[the remark this answers is not in the document]
Comment (geoffreyQiu) [reply examples/commons/datasets/hstu_batch.py:248]: Use `dynamic_shapes._IntWrapper` for now.

File: corelib/dynamicemb/src/table_operation/expand_table_ids_torch_binding.cu [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+    if (table_offsets_in_feature.has_value() &&
+        table_offsets_in_feature.value().defined()) {
+        TORCH_CHECK(
+            table_offsets_in_feature.value().is_cuda(),
+            "INFERENCE_EMB::expand_table_ids expects CUDA table_offsets_in_feature when provided."
+        );
+    }
+    return expand_table_ids_cuda(offsets, table_offsets_in_feature, num_tables,
+                                 local_batch_size, num_elements);
+}
```
[the remark this answers is not in the document]
Comment (geoffreyQiu) [reply corelib/dynamicemb/src/table_operation/expand_table_ids_torch_binding.cu:25]: Keep `indices` for dynamic shape deduction for now.

**Review by shijieliu — APPROVED** [review]
